File: SASMopubBannerCustomEvent.m
//
//  SASMopubBannerCustomEvent.m
//  SmartAdServer
//
//  Created by Thomas Geley on 21/12/2016.
//  Copyright © 2019 Smart AdServer. All rights reserved.
//

#import "MPLogging.h"
#import "MoPub.h"
#import "SASMopubBannerCustomEvent.h"
#import "SASMopubUtils.h"
#import <SASDisplayKit/SASDisplayKit.h>

NS_ASSUME_NONNULL_BEGIN

@interface SASMopubBannerCustomEvent () <SASBannerViewDelegate>

@property (nonatomic, strong, nullable) SASBannerView *bannerView;

@end

@implementation SASMopubBannerCustomEvent
@dynamic delegate;
@dynamic localExtras;

#pragma mark - Object Lifecycle

- (void)dealloc {
    [self destroyAdView];
}

- (void)destroyAdView {
    self.bannerView.delegate = nil;
    self.bannerView = nil;
}

#pragma mark - MPInlineAdAdapter Override

- (BOOL)enableAutomaticImpressionAndClickTracking {
    return YES;
}

- (void)requestAdWithSize:(CGSize)size adapterInfo:(NSDictionary *)info adMarkup:(NSString * _Nullable)adMarkup {
    
    // Reset banner view
    [self destroyAdView];
    
    // Extracting placement from custom event info
    NSError *error = nil;
    SASAdPlacement *adPlacement = [SASMopubUtils adPlacementWithCustomEventInfo:info error:&error];
    
    if (adPlacement == nil) {
        // Failing if custom info are invalid
        [self.delegate inlineAdAdapter:self didFailToLoadAdWithError:error];
        return;
    }
    
    // Creating a banner instance
    CGRect frame = CGRectMake(0, 0, size.width, size.height);
    self.bannerView = [[SASBannerView alloc] initWithFrame:frame];
    self.bannerView.delegate = self;
    self.bannerView.modalParentViewController = [self.delegate inlineAdAdapterViewControllerForPresentingModalView:self];
    self.bannerView.autoresizingMask = UIViewAutoresizingFlexibleWidth;
    
    // Loading ad from ad placement
    [self.bannerView loadWithPlacement:adPlacement];
    
}

#pragma mark - Banner delegate

- (void)bannerViewDidLoad:(SASBannerView *)bannerView {
    MPLogInfo(@"Smart Banner did load");
    [self.delegate inlineAdAdapter:self didLoadAdWithAdView:self.bannerView];
}

- (void)bannerView:(SASBannerView *)bannerView didFailToLoadWithError:(NSError *)error {
    MPLogInfo(@"Smart Banner failed to load with error: %@", error.localizedDescription);
    [self.delegate inlineAdAdapter:self didFailToLoadAdWithError:error];
}

- (void)bannerViewWillPresentModalView:(SASBannerView *)bannerView {
    MPLogInfo(@"Smart Banner will present modal");
    [self.delegate inlineAdAdapterWillBeginUserAction:self];
}

- (void)bannerViewWillDismissModalView:(SASBannerView *)bannerView {
    MPLogInfo(@"Smart Banner did dismiss modal");
    [self.delegate inlineAdAdapterDidEndUserAction:self];
}

@end

NS_ASSUME_NONNULL_END

File: SASMopubBannerCustomEvent.swift
//
//  SASMopubBannerCustomEvent.swift
//  SASMopubBannerCustomEvent
//
//  Created by Loïc GIRON DIT METAZ on 01/09/2021.
//

import Foundation
import SASDisplayKit
import MoPubSDK

@objc(SASMopubBannerCustomEvent)
class SASMopubBannerCustomEvent: MPInlineAdAdapter {
    
    private var bannerView: SASBannerView? = nil
    
    override func requestAd(with size: CGSize, adapterInfo info: [AnyHashable : Any], adMarkup: String?) {
        do {
            // Extracting placement from custom event info
            let adPlacement = try SASMopubUtils.adPlacement(customEventInfo: info)
            
            // Creating a banner instance
            let rect = CGRect(x: 0.0, y: 0.0, width: size.width, height: size.height)
            bannerView = SASBannerView(frame: rect)
            bannerView?.delegate = self
            bannerView?.modalParentViewController = delegate?.inlineAdAdapterViewController(forPresentingModalView: self)
            bannerView?.autoresizingMask = [.flexibleWidth]
            
            // Loading ad from ad placement
            bannerView?.load(with: adPlacement)
        } catch {
            // Failing if custom info are invalid
            delegate?.inlineAdAdapter(self, didFailToLoadAdWithError: error)
        }
    }
    
    override func enableAutomaticImpressionAndClickTracking() -> Bool {
        return true
    }
    
}

extension SASMopubBannerCustomEvent: SASBannerViewDelegate {
    
    func bannerViewDidLoad(_ bannerView: SASBannerView) {
        if let bannerView = self.bannerView {
            delegate?.inlineAdAdapter(self, didLoadAdWithAdView: bannerView)
        }
    }
    
    func bannerView(_ bannerView: SASBannerView, didFailToLoadWithError error: Error) {
        delegate?.inlineAdAdapter(self, didFailToLoadAdWithError: error)
    }
    
    func bannerViewWillPresentModalView(_ bannerView: SASBannerView) {
        delegate?.inlineAdAdapterWillBeginUserAction(self)
    }
    
    func bannerViewWillDismissModalView(_ bannerView: SASBannerView) {
        delegate?.inlineAdAdapterDidEndUserAction(self)
    }
    
}
